### PR TITLE
Expand watchOS support for GoogleDataTransport

### PR DESCRIPTION
**Pull request by doudounan**
- Add watchOS support for GoogleDataTransport with specifying notification center and WKExtensionDelegate for watchOS application, handling background tasks and network reachability flags on watchOS.
- Add independent watchOS test app for GoogleDataTransport and GoogleDataTransportCCTSupport.
- Add companion watchOS test app and notification service extension for GoogleDataTransportCCTSupport.

### CODE REVIEW

File: GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m [hunk 1]
```diff
@@ -245,6 +257,14 @@ @interface GDTCORApplication ()
 
 @implementation GDTCORApplication
 
+/** A dispatch queue on which all task semaphore will populate and remove from
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m:260]: Nit: semaphores

File: GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m [hunk 2]
```diff
@@ -245,6 +257,14 @@ @interface GDTCORApplication ()
...
+ * gBackgroundIdentifierToSemaphoreMap.
+ */
+static dispatch_queue_t gSemaphoreQueue;
+
+/** For mapping backgroundIdentifier to task semaphore. */
+static NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *gBackgroundIdentifierToSemaphoreMap;
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m:266]: Should this be wrapped in `#if TARGET_OS_WATCH`?

File: GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m [hunk 3]
```diff
@@ -310,67 +345,172 @@ - (instancetype)init {
...
   }
   return self;
 }
 
+/** Generates and maps a unique background identifier to the given semaphore.
+ *
+ * @param semaphore The semaphore to map.
+ * @return A unique GDTCORBackgroundIdentifier mapped to the given semaphore.
+ */
++ (GDTCORBackgroundIdentifier)createAndMapBackgroundIdentifierToSemaphore:
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m:381]: Should the next two functions be wrapped in `#if TARGET_OS_WATCH`

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORPlatform.h [hunk 4]
```diff
@@ -66,20 +69,47 @@ typedef NS_ENUM(NSInteger, GDTCORNetworkMobileSubtype) {
...
+typedef SCNetworkReachabilityFlags GDTCORNetworkReachabilityFlags;
+
+/** Define SCNetworkReachabilityRef as GDTCORNetworkReachabilityRef on non-watchOS. */
+typedef SCNetworkReachabilityRef GDTCORNetworkReachabilityRef;
+
+#else
+/** The different possible reachabilityFlags option on watchOS. */
+typedef NS_OPTIONS(uint32_t, GDTCORNetworkReachabilityFlags) {
+  kGDTCORNetworkReachabilityFlagsReachable = 1 << 1,
+  // TODO(doudounan): Add more options on watchOS if needed.
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/Public/GDTCORPlatform.h:83]: How will we know if more are needed?

File: GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m (hunk 2 above)
Comment (doudounan) [reply GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m:266]: Even though those two variable does not initialize in other platforms, wrapped in` #if TARGET_OS_WATCH` should be good.

File: GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m (hunk 3 above)
Comment (doudounan) [reply GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m:381]: Same as above.

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORPlatform.h (hunk 4 above)
Comment (doudounan) [reply GoogleDataTransport/GDTCORLibrary/Public/GDTCORPlatform.h:83]: 'Needed' is not explicit in this situation. I will change it into 'if watchOS network connection information relative apis available in the future.

File: GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m (hunk 1 above)
Comment (doudounan) [reply GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m]: Fixed

File: GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m (hunk 2 above)
Comment (mikehaney24) [reply GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m:267]: LGTM

File: GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m (hunk 3 above)
Comment (mikehaney24) [reply GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m:384]: LGTM

**Pull request merged**